eq3bt/connection.py

```diff
@@ -15,46 +17,83 @@
         """Initialize the connection."""
         btle.DefaultDelegate.__init__(self)
 
+        self._ifaces = self.get_hci_ifaces()
+        self._iface_idx = 0
+
         self._conn = None
         self._mac = mac
         self._callbacks = {}
+        self._keep_connected = keep_connected
+    
+    def next_iface(self):
+        self._nr_conn_errors += 1
+        self._iface_idx = (self._iface_idx + 1) % len(self._ifaces)
+        if self._nr_conn_errors >= len(self._ifaces)*2:
+            return False
+        return True
+
+    def get_hci_ifaces(self):
+        iface_list = []
+        bus = dbus.SystemBus()
+        manager = dbus.Interface(bus.get_object("org.bluez", "/"), "org.freedesktop.DBus.ObjectManager")
+        objects = manager.GetManagedObjects()
+        for path, interfaces in objects.items():
+            adapter = interfaces.get("org.bluez.Adapter1")
+            if adapter is None:
+                continue
+            iface_list.append(re.search(r'\d+$', path)[0])
+        return iface_list
 
     def __enter__(self):
         """
         Context manager __enter__ for connecting the device
         :rtype: btle.Peripheral
         :return:
         """
-        self._conn = btle.Peripheral()
-        self._conn.withDelegate(self)
-        _LOGGER.debug("Trying to connect to %s", self._mac)
         try:
-            self._conn.connect(self._mac)
-        except btle.BTLEException as ex:
-            _LOGGER.debug("Unable to connect to the device %s, retrying: %s", self._mac, ex)
-            try:
-                self._conn.connect(self._mac)
-            except Exception as ex2:
-                _LOGGER.debug("Second connection try to %s failed: %s", self._mac, ex2)
-                raise
+            conn_state = self._conn.getState()
+        except:
+            self._conn = None
+
+        if self._conn is None or conn_state != "conn":
+            self._conn = btle.Peripheral()
+            self._conn.withDelegate(self)
+            self._nr_conn_errors = 0
+            _LOGGER.debug("Trying to connect to %s", self._mac)
+            while True:
+                # try to connect with all ifaces
+                try:
+                    self._conn.connect(self._mac, iface=self._ifaces[self._iface_idx])
+                    break
+                except btle.BTLEException as ex:
+                    _LOGGER.debug("Unable to connect to the device %s using iface %s, retrying: %s", self._mac, self._ifaces[self._iface_idx], ex)
+                    try:
+                        self._conn.connect(self._mac, iface=self._ifaces[self._iface_idx])
+                        break
+                    except Exception as ex2:
+                        _LOGGER.debug("Second connection try to %s using ifaces %s failed: %s", self._mac, self._ifaces[self._iface_idx], ex2)
+                        if self.next_iface() is False:
+                            # tried all ifaces, raise exception
+                            raise
```

Comment:
This structure feels very error-prone, I think the correct way would be refactoring the connection code away from `__enter__` (it's not really exposed anywhere), and create separate `connect()` method that does the connection using the given iface.

Reply:
That makes sense. I'll refactor and add a `connect()` method with iface as parameter.